File: packages/react-native-nitro-thermal/android/src/main/java/com/margelo/nitro/nitrothermal/HybridThermalMonitor.kt
package com.margelo.nitro.nitrothermal

import android.content.Context
import android.os.PowerManager
import androidx.annotation.Keep
import com.facebook.proguard.annotations.DoNotStrip
import com.margelo.nitro.NitroModules

@Keep
@DoNotStrip
class HybridThermalMonitor : HybridThermalMonitorSpec() {

    private val context: Context
        get() = NitroModules.applicationContext
            ?: throw IllegalStateException("React context is not available")

    private val powerManager: PowerManager
        get() = context.getSystemService(Context.POWER_SERVICE) as PowerManager

    // MARK: - Get current thermal state

    override fun getCurrentThermalState(): ThermalState {
        return mapAndroidThermalStatus(powerManager.currentThermalStatus)
    }

    // MARK: - Listener

    override fun addThermalStateListener(
        listener: (state: ThermalState) -> Unit
    ): () -> Unit {
        val thermalListener = PowerManager.OnThermalStatusChangedListener { status ->
            listener(mapAndroidThermalStatus(status))
        }

        powerManager.addThermalStatusListener(thermalListener)

        return {
            powerManager.removeThermalStatusListener(thermalListener)
        }
    }

    // MARK: - Helpers

    private fun mapAndroidThermalStatus(status: Int): ThermalState {
        return when (status) {
            PowerManager.THERMAL_STATUS_NONE -> ThermalState.NOMINAL
            PowerManager.THERMAL_STATUS_LIGHT -> ThermalState.FAIR
            PowerManager.THERMAL_STATUS_MODERATE -> ThermalState.FAIR
            PowerManager.THERMAL_STATUS_SEVERE -> ThermalState.SERIOUS
            PowerManager.THERMAL_STATUS_CRITICAL -> ThermalState.CRITICAL
            PowerManager.THERMAL_STATUS_EMERGENCY -> ThermalState.CRITICAL
            PowerManager.THERMAL_STATUS_SHUTDOWN -> ThermalState.CRITICAL
            else -> ThermalState.UNKNOWN
        }
    }
}

File: packages/react-native-nitro-thermal/ios/HybridThermalMonitor.swift
import Foundation
import NitroModules

class HybridThermalMonitor: HybridThermalMonitorSpec {

  // MARK: - Get current thermal state

  func getCurrentThermalState() throws -> ThermalState {
    return mapProcessInfoThermalState(ProcessInfo.processInfo.thermalState)
  }

  // MARK: - Listener

  func addThermalStateListener(
    listener: @escaping (_ state: ThermalState) -> Void
  ) throws -> () -> Void {
    let observer = NotificationCenter.default.addObserver(
      forName: ProcessInfo.thermalStateDidChangeNotification,
      object: nil,
      queue: .main
    ) { _ in
      let mapped = self.mapProcessInfoThermalState(
        ProcessInfo.processInfo.thermalState
      )
      listener(mapped)
    }

    // Emit current state immediately to match Android behavior
    listener(mapProcessInfoThermalState(ProcessInfo.processInfo.thermalState))

    // Return a cleanup function that removes the observer
    return {
      NotificationCenter.default.removeObserver(observer)
    }
  }

  // MARK: - Helpers

  private func mapProcessInfoThermalState(
    _ state: ProcessInfo.ThermalState
  ) -> ThermalState {
    switch state {
    case .nominal:
      return .nominal
    case .fair:
      return .fair
    case .serious:
      return .serious
    case .critical:
      return .critical
    @unknown default:
      return .unknown
    }
  }
}
